src/methods/join.js

```diff
@@ -301,17 +299,37 @@ export class JoinImplementation {
       let nae = new Effects(na, ae.generator, ae.modifiedBindings, ae.modifiedProperties, ae.createdObjects);
       return new ForkedAbruptCompletion(realm, pnc.joinCondition, pncc, pnc.consequentEffects, na, nae);
     } else {
+      let nc, nce;
+      if (pncc instanceof PossiblyNormalCompletion) {
+        nc = this.replacePossiblyNormalCompletionWithForkedAbruptCompletion(realm, pncc, ac, e);
+        let ce = pnc.consequentEffects;
+        nce = new Effects(nc, ce.generator, ce.modifiedBindings, ce.modifiedProperties, ce.createdObjects);
+      } else {
+        invariant(pncc instanceof SimpleNormalCompletion);
+        nc = ac;
+        nce = realm.composeEffects(pnc.consequentEffects, e);
+      }
       let pnca = pnc.alternate;
-      invariant(pnca instanceof AbruptCompletion);
-      e = realm.composeEffects(pnc.consequentEffects, e);
-      if (pnc.consequent instanceof SimpleNormalCompletion) {
-        return new ForkedAbruptCompletion(realm, pnc.joinCondition, ac, e, pnca, pnc.alternateEffects);
+      let na, nae;
+      // TODO (hermanv) if we use e as is, it ends up being applied twice when the join of the normal
+      // path is applied to the current state. Follow up with a PR that allows Effects to get deeply cloned
```

Comment:
What does "deeply cloning" do to the generators (and generator entries) in it?

Reply:
It would have to clone all that. On second thoughts, though, I would like to do something that does not amount to tail duplication.

Reply:
Can we ensure the design is such that no generators get cloned? If that's not conceivable right now, let's that about it.

Reply:
I am busy conceiving it.